docs/syntax-property-access-modifiers.md

```diff
@@ -0,0 +1,195 @@
+
+```lua
+type Reader = { read read: () -> number }
+type Writer = { read write: (number) -> () }
+```
+
+It is important to consider that this will be an issue for any keywords we might
+choose unless we were to take the step of picking something unlikely to be a
+useful property name. (ie something weird looking and ugly)
+
+Lastly, mixed indexers are very awkward both in syntax and semantics.
+
+## Alternatives
+
+The design space for syntax includes:
+
+* Names, symbols or attributes?
+* Modifier position?
+* How to parse or serialize properties with differing read- and write-types?
+
+### Names, symbols or attributes?
+
+We could use names for modifiers, such as
+
+* `get` and `set`
+* `const` and `mut`
+* `read` and `write`
+* `readonly` and `writeonly`
+
+One issue is that these are all valid identifiers, so if we want
+backward compatbility, they cannot be made keywords. This presents
+issues with code that uses the chosen names as type or property names,
+for instance:
+
+```lua
+  type set = { [any] : bool }
+  type ugh = { get set : set }
+```
+
+We could use symbols, for example
+
+* `+` and `-`
+* (Are there other obvious pairs of symbols?)
+
+We could use attributes, for example
+
+* `@read` and `@write`
+* ... as per "names for modifiers", only prefixed by `@` ...
+
+These both have the advantage of being unambiguous and easier to parse. Symbols
+are terser, whch is both good and bad.
+
+We decided not to use glyphs because they are more difficult to understand,
+don't contribute very much, and aren't very stylistically consistent with other
+Luau syntax.
+
+### Modifier position?
+
+For attributes, the position is given by the syntax of attributes, for example:
+
+```lua
+  type Vector2 = { @read x: number, @read y : Number }
+```
+
+For the other proposals, there are four possibilities, depending on whether the
+modifier is west-coast or east-coast, and whether it modifies the propertry name
+or the type:
+
+```lua
+  type Vector2 = { read x : number, read y : number }
+  type Vector2 = { x read : number, y read : number }
+  type Vector2 = { x : read number, y : read number }
+  type Vector2 = { x : number read, y : number read }
+```
+
+The east-coast options are not easy-to-read with names, but are
+easier with symbols, especially since `T?` is already postfix, for
+example
+
+```lua
+  type Foo = { p: number?+ }
+```
+
+### How to parse or serialize properties with differing read- and write-types?
+
+One corner case is that type inference may deduce different read- and
+write-types, which need to be presented to the user. For example the
+read-type of `x` is `Animal` but its write-type is `Dog` in the principal type of:
+
+```lua
+   function f(x)
+     let a: Animal = x.pet
+     x.pet = Dog.new()
+     return a
+   end
+```
+
+If we are adding the modifier to the property name, we can repeat the name, for example
+
+```lua
+  x : { read pet : Animal, write pet : Dog }
+```
```

Comment:
While we have types that are written out in a way that can't be parsed today (ToStringResult::invalid), I think it's reasonable to try to print out in a way that can be put back into the parser.